src/Scope.hs

```diff
@@ -1,14 +1,31 @@
 {-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE CPP                #-}
 {-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
 module Scope where
 
 import Data.Data (Data)
 import Control.Lens
 
+#ifdef KDEBUG
 import Data.Text (Text)
+#else
+import Util.Key
+#endif
+
 
 -- Int should be enough for now - consider bumping to something like int64
+#ifndef KDEBUG
+newtype Scope = Scope { scopeNum :: Int}
+  deriving newtype (Eq, Ord, Show, HasKey)
+  deriving stock Data
+#else
+-- For a debug build Scope keeps a blob of text for debugging the expander
+-- output. This will have an impact of the performance of the interpreter so it
+-- won't be useful for performance issues
 data Scope = Scope { scopeNum :: Int, scopePurpose :: Text }
   deriving (Data, Eq, Ord, Show)
```

Comment:
Since we know that any two `Scope`s with the same `scopeNum` will have the same `scopePurpose`, would it make sense for the `Eq` and `Ord` instances to only look at `scopeNum`?

Reply:
Yes this would make perfect sense. Now that I know where the leaks are coming from I can experiment with adding a `scopePurpose` Store to the `Expander` environment. The only catch is that because that we would always have to be careful around that Store because if it is never forced then it'll still build up thunks even though Store is a strict IntMap

Reply:
I don't love the CPP, but I also don't necessarily hate it, especially because from our conversations the `scopePurpose` seemed to be used to debug correctness not performance so a debug build would make sense in that case.

Reply:
> The only catch is that because that we would always have to be careful around that Store because if it is never forced then it'll still build up thunks even though Store is a strict IntMap

The `scopePurpose` is only forced if it is part of an error message. Are saying that while the strict IntMap makes sure the force the thunk which produces the Text, we still need to remember to force the Store itself whenever we update the `Expander` field which holds the Store, otherwise nothing is going to force the thunk which calculates this update? Wouldn't it be sufficient to make that `Expander` field strict, using a `!`?